src/Optimizer/Transformer/RewriteAmpUrls.php

```diff
@@ -0,0 +1,356 @@
+                &&
+                $this->usesAmpCacheUrl($href)
+            ) {
+                $newUrl = $this->replaceUrl($href, $host);
+                $node->setAttribute(Attribute::HREF, $newUrl);
+                $preloadNodes[] = $this->createPreload($document, $newUrl, Tag::STYLE);
+            } elseif (
+                $node->tagName === Tag::LINK
+                &&
+                $node->getAttribute(Attribute::REL) === Attribute::REL_PRELOAD
+                &&
+                $this->usesAmpCacheUrl($href)
+            ) {
+                if ($usesEsm && $this->shouldPreload($href)) {
+                    // Only preload .mjs runtime in ESM mode.
+                    $node->parentNode->removeChild($node);
+                } else {
+                    $node->setAttribute(Attribute::HREF, $this->replaceUrl($href, $host));
+                }
+            }
+
+            $node = $node->nextSibling;
+        }
+
+        return $preloadNodes;
+    }
+
+    /**
+     * Check if a given URL uses the AMP Cache.
+     *
+     * @param string $url Url to check.
+     * @return bool Whether the provided URL uses the AMP Cache.
+     */
+    private function usesAmpCacheUrl($url)
+    {
+        if (! $url) {
+            return false;
+        }
+
+        return strpos($url, Amp::CACHE_HOST) === 0;
+    }
+
+    /**
+     * Replace URL root with provided host.
+     *
+     * @param string $url  Url to replace.
+     * @param string $host Host to use.
+     * @return string Adapted URL.
+     */
+    private function replaceUrl($url, $host)
+    {
+        return str_replace(Amp::CACHE_HOST, $host, $url);
+    }
+
+    /**
+     * Replace <script> elements with their ES module counterparts.
+     *
+     * @param Document $document   Document to add the ES module scripts to.
+     * @param Element  $scriptNode Script element to replace.
+     * @return Element|null Preload element that was added, or null if none was required or an error occurred.
+     */
+    private function addEsm(Document $document, Element $scriptNode)
+    {
+        $preloadNode  = null;
+        $scriptUrl    = $scriptNode->getAttribute(Attribute::SRC);
+        $esmScriptUrl = preg_replace('/\.js$/', '.mjs', $scriptUrl);
+
+        if ($this->shouldPreload($scriptUrl)) {
+            $preloadNode = $document->createElement(Tag::LINK);
+            $preloadNode->setAttribute(Attribute::AS_, Tag::SCRIPT);
+            $preloadNode->setAttribute(Attribute::CROSSORIGIN, Attribute::CROSSORIGIN_ANONYMOUS);
+            $preloadNode->setAttribute(Attribute::HREF, $esmScriptUrl);
+            $preloadNode->setAttribute(Attribute::REL, Attribute::REL_MODULEPRELOAD);
+        }
+
+        $nomoduleNode = $document->createElement(Tag::SCRIPT);
+        $nomoduleNode->addBooleanAttribute(Attribute::ASYNC);
+        $nomoduleNode->addBooleanAttribute(Attribute::NOMODULE);
+        $nomoduleNode->setAttribute(Attribute::SRC, $scriptUrl);
```

Comment:
Shouldn't the `script[nomodule]` get `crossorigin=anonymous` too? 

```suggestion
        $nomoduleNode->setAttribute(Attribute::SRC, $scriptUrl);
        $nomoduleNode->setAttribute(Attribute::CROSSORIGIN, Attribute::CROSSORIGIN_ANONYMOUS);
```

Reply:
Depends on https://github.com/ampproject/amp-toolbox/pull/1194

Reply:
Merging for now (to get tests to pass in other PRs, and creating a separate issue for this.

Reply:
=> https://github.com/ampproject/amp-toolbox-php/pull/140